src/models/room.rs

```diff
@@ -0,0 +1,298 @@
+//
+// Unless required by applicable law or agreed to in writing, software
+// distributed under the License is distributed on an "AS IS" BASIS,
+// WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+// See the License for the specific language governing permissions and
+// limitations under the License.
+
+use std::collections::HashMap;
+use std::sync::{Arc, RwLock};
+
+use super::{RoomMember, User, UserId};
+
+use crate::events::collections::all::{RoomEvent, StateEvent};
+use crate::events::room::{
+    aliases::AliasesEvent,
+    canonical_alias::CanonicalAliasEvent,
+    member::{MemberEvent, MembershipChange},
+    name::NameEvent,
+    power_levels::PowerLevelsEvent,
+};
+use crate::events::{
+    presence::{PresenceEvent, PresenceEventContent},
+    EventResult,
+};
+use crate::identifiers::RoomAliasId;
+use crate::session::Session;
+
+use js_int::UInt;
+
+#[cfg(feature = "encryption")]
+use tokio::sync::Mutex;
+
+#[cfg(feature = "encryption")]
+use crate::crypto::{OlmMachine, OneTimeKeys};
+#[cfg(feature = "encryption")]
+use ruma_client_api::r0::keys::{upload_keys::Response as KeysUploadResponse, DeviceKeys};
+
+#[derive(Debug, Default)]
+/// `RoomName` allows the calculation of a text room name.
+pub struct RoomName {
+    /// The displayed name of the room.
+    name: Option<String>,
+    /// The canonical alias of the room ex. `#room-name:example.com` and port number.
+    canonical_alias: Option<RoomAliasId>,
+    /// List of `RoomAliasId`s the room has been given.
+    aliases: Vec<RoomAliasId>,
+}
+
+#[derive(Debug)]
+/// A Matrix rooom.
+pub struct Room {
+    /// The unique id of the room.
+    pub room_id: String,
+    /// The name of the room, clients use this to represent a room.
+    pub room_name: RoomName,
+    /// The mxid of our own user.
+    pub own_user_id: UserId,
+    /// The mxid of the room creator.
+    pub creator: Option<UserId>,
+    /// The map of room members.
+    pub members: HashMap<UserId, RoomMember>,
+    /// A list of users that are currently typing.
+    pub typing_users: Vec<UserId>,
+    /// A flag indicating if the room is encrypted.
+    pub encrypted: bool,
+    /// Number of unread notifications with highlight flag set.
+    pub unread_highlight: Option<UInt>,
+    /// Number of unread notifications.
+    pub unread_notifications: Option<UInt>,
+}
+
+impl RoomName {
+    pub fn push_alias(&mut self, alias: RoomAliasId) -> bool {
+        self.aliases.push(alias);
+        true
+    }
+
+    pub fn set_canonical(&mut self, alias: RoomAliasId) -> bool {
+        self.canonical_alias = Some(alias);
+        true
+    }
+
+    pub fn set_name(&mut self, name: &str) -> bool {
+        self.name = Some(name.to_string());
+        true
+    }
+
+    pub fn calculate_name(&self, room_id: &str, members: &HashMap<UserId, RoomMember>) -> String {
+        // https://github.com/matrix-org/matrix-js-sdk/blob/33941eb37bffe41958ba9887fc8070dfb1a0ee76/src/models/room.js#L1823
```

Comment:
We should rather link to the spec https://matrix.org/docs/spec/client_server/latest#calculating-the-display-name-for-a-room.